maitake/src/wait/queue.rs

```diff
@@ -0,0 +1,790 @@
+/// [wake]: WaitQueue::wake
+/// [wake_all]: WaitQueue::wake_all
+/// [`UnsafeCell`]: core::cell::UnsafeCell
+/// [ilist]: cordyceps::List
+/// [intrusive]: https://fuchsia.dev/fuchsia-src/development/languages/c-cpp/fbl_containers_guide/introduction
+/// [2]: https://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue
+#[derive(Debug)]
+pub struct WaitQueue {
+    /// The wait queue's state variable.
+    state: CachePadded<AtomicUsize>,
+
+    /// The linked list of waiters.
+    ///
+    /// # Safety
+    ///
+    /// This is protected by a mutex; the mutex *must* be acquired when
+    /// manipulating the linked list, OR when manipulating waiter nodes that may
+    /// be linked into the list. If a node is known to not be linked, it is safe
+    /// to modify that node (such as by waking the stored [`Waker`]) without
+    /// holding the lock; otherwise, it may be modified through the list, so the
+    /// lock must be held when modifying the
+    /// node.
+    ///
+    /// A spinlock (from `mycelium_util`) is used here, in order to support
+    /// `no_std` platforms; when running `loom` tests, a `loom` mutex is used
+    /// instead to simulate the spinlock, because loom doesn't play nice with
+    /// real spinlocks.
+    queue: Mutex<List<Waiter>>,
+}
+
+/// Future returned from [`WaitQueue::wait()`].
+///
+/// This future is fused, so once it has completed, any future calls to poll
+/// will immediately return [`Poll::Ready`].
+#[derive(Debug)]
+#[pin_project(PinnedDrop)]
+#[must_use = "futures do nothing unless `.await`ed or `poll`ed"]
+pub struct Wait<'a> {
+    /// The [`WaitQueue`] being waited on from.
+    queue: &'a WaitQueue,
+
+    /// Entry in the wait queue linked list.
+    #[pin]
+    waiter: Waiter,
+}
+
+/// The state of a [`Waiter`] node in a [`WaitQueue`].
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum WaitState {
+    /// The waiter has not yet been enqueued.
+    ///
+    /// The number of times [`WaitQueue::wake_all`] has been called is stored
+    /// when the node is created, in order to determine whether it was woken by
+    /// a stored wakeup when enqueueing.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Start(usize),
+
+    /// The waiter is waiting.
+    ///
+    /// When in this state, the node **is** part of the linked list. If the
+    /// node is dropped in this state, it **must** be removed from the list
+    /// before dropping it. Failure to ensure this will result in dangling
+    /// pointers in the linked list!
+    Waiting,
+
+    /// The waiter has been woken.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Woken,
+}
+
+/// A waiter node which may be linked into a wait queue.
+#[derive(Debug)]
+#[repr(C)]
+#[pin_project]
+struct Waiter {
```

Comment:
i don't really think this is a `Header`; in the case of the task header, the naming is because the header is a small amount of data stored before the task's "body" (the future/output). in this case, the `Waiter` type is the entirety of the data in the linked list...